core/controller/src/main/scala/whisk/core/controller/Actions.scala

```diff
@@ -781,6 +796,318 @@ trait WhiskActionsApi extends WhiskCollectionAPI {
+     * @param env the merged parameters from the package/reference if any TODO: check ignoring these params
+     * @param payload the dynamic arguments for the activation
+     * @param blocking true iff this is a blocking invoke
+     * @param topmost true iff this is the topmost sequence invoked directly through the api (not indirectly through a sequence)
+     * @param components the actions in the sequence
+     * @param cause the id of the activation that caused this sequence (defined only for inner sequences and None for topmost sequences)
+     * @param atomicActionsCount the dynamic atomic action count observed so far since the start of invocation of the topmost sequence(0 if topmost)
+     * @param transid a transaction id for logging
+     * @return a future of type (ActivationId, Some(WhiskActivation), atomicActionsCount) if blocking; else (ActivationId, None, 0)
+     */
+    private def invokeSequence(
+            user: Identity,
+            action: WhiskAction,
+            //env: Option[Parameters],
+            payload: Option[JsObject],
+            blocking: Boolean,
+            topmost: Boolean,
+            components: Vector[FullyQualifiedEntityName],
+            cause: Option[ActivationId],
+            atomicActionsCount: Int)(implicit transid: TransactionId): Future[(ActivationId, Option[WhiskActivation], Int)] = {
+        // create new activation id that corresponds to the sequence
+        val seqActivationId = activationIdFactory.make()
+        info(this, s"Invoking sequence $action topmost $topmost activationid '$seqActivationId'")
+        val start = Instant.now(Clock.systemUTC())
+        var seqActivation: Option[WhiskActivation] = None
+        // the cause for the component activations is the current sequence
+        val futureWskActivations = invokeSequenceComponents(user, action, seqActivationId, payload, components, cause = Some(seqActivationId), atomicActionsCount)
+        val futureSeqResult = Future.sequence(futureWskActivations)
+        val response: Future[(ActivationId, Option[WhiskActivation], Int)] =
+            if (topmost) { // need to deal with blocking and closing connection
+                if (blocking) {
+                    val timeout = maxWaitForBlockingActivation + blockingInvokeGrace
+                    val futureSeqResultTimeout = futureSeqResult withTimeout(timeout, new BlockingInvokeTimeout(seqActivationId))
+                    // if the future fails with a timeout, the failure is dealt with at the caller level
+                    futureSeqResultTimeout map { wskActivationTuples =>
+                        val wskActivationEithers = wskActivationTuples.map(_._1)
+                        // the execution of the sequence was successful, return the result
+                        val end = Instant.now(Clock.systemUTC())
+                        seqActivation = Some(makeSequenceActivation(user, action, seqActivationId, wskActivationEithers, topmost, cause, start, end))
```

Comment:
this flow for creating the seqActivation is repeated three times - the only difference is the `end` marker. 
since the `end` marker is not that important for the sequence, is it really worth the code duplication?